lib/vm/src/memory.rs

```diff
@@ -616,4 +643,15 @@ where
 
     /// Copies this memory to a new memory
     fn duplicate(&mut self) -> Result<Box<dyn LinearMemory + 'static>, MemoryError>;
+
+    /// Add current thread to the waiter hash, and wait until notified or timout.
+    /// Return 0 if the waiter has been notified, 2 if the timeout occured, or None if en error happened
+    fn do_wait(&mut self, _dst: NotifyLocation, _timeout: Option<Duration>) -> Option<u32> {
```

Comment:
We should return a `Result<_, Trap` here to allow being specific about the trap. 

See some related comments on this.

Reply:
This will be especially important when we allow customizing this behaviour via a trait.

Reply:
But what Trap should be returned for the non-implemented path?

Reply:
I change to return a Result with custom error instead @theduke . Is that better now?